codegen: Allow string arguments to have defaults

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45668 Use asymmetric mkldnn convolutions for padding='same' convolutions
* #45667 Add padding='same' mode to conv{1,2,3}d
* #45666 codegen: Resolve overload ambiguities created by defaulted arguments
* **#45665 codegen: Allow string arguments to have defaults**

Fixes #43944

Note that the codegen doesn't use a proper parser so, in the same way as with lists, the string `, ` cannot appear in defaults or it will be interpreted as a splitting point between arguments.

Differential Revision: [D24141835](https://our.internmc.facebook.com/intern/diff/D24141835)

peterbell10 committed:
codegen: Allow string arguments to have defaults

[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Allow string arguments to have defaults"


Fixes #43944

Note that the codegen doesn't use us a proper parser so, in the same way as with lists, the string `, ` cannot appear in defaults or it will be interpreted as a splitting point between arguments.

[ghstack-poisoned]

ezyang commented on tools/codegen/api/cpp.py:
```diff
@@ -158,6 +158,26 @@ def returns_type(rs: Sequence[Return]) -> str:
+        if len(d) >= 2 and d[0] == "'" and d[-1] == "'":
+            s = ''
+            i = 1
+            while i + 1 < len(d):
```
This chunk of code is what I would expect to see in C, but not in Python. Manual indexing is verbose, and holds up poorly when you need to add extensions in the future. In Python, you should just parse the type into an actual string, and then escape it for double quotes (which would just be a simple `replace('"', '\\"')`)

ezyang replied:
(Parsing can be done with `ast.literal_eval`, if we define the permissible string format to be all of Python (which should be the case, given that we support single quotes)

ezyang replied:
Or, better yet, we can disallow single-quoted strings and then define the string format to be exactly JSON string format.

peterbell10 replied:
I am only allowing single quotes because the JIT frontend also allows it. I can remove this if not matching the JIT frontend exactly is okay.

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -503,6 +503,62 @@ static inline std::vector<int64_t> parse_intlist_args(const std::string& s, int6
 }
 
+// Parse a string literal to remove quotes and escape sequences
+static std::string parse_string_literal(std::string str) {
```
You shouldn't be taking `std::string` by value here as it induces a copy. Use `c10::string_view`

peterbell10 replied:
Fixed

ezyang reviewed (approved):
OK

peterbell10 committed:
Update on "codegen: Allow string arguments to have defaults"


Fixes #43944

Note that the codegen doesn't use a proper parser so, in the same way as with lists, the string `, ` cannot appear in defaults or it will be interpreted as a splitting point between arguments.

[ghstack-poisoned]

peterbell10 commented on tools/codegen/gen.py:
```diff
@@ -773,9 +781,12 @@ def compute_registration_declarations(f: NativeFunction) -> str:
+        'dispatch': str(f.dispatch is not None),
+        'math': str(f.dispatch is not None and 'Math' in f.dispatch)
+    }
+    return f"""{returns_type} {name}({args_str}); // {json.dumps(comment_data)}
```
I think this is the source of the XLA failures. The old code left the schema un-escaped which failed in JSON parsing. A bit confusing since the error message doesn't mention the JSON at all.

peterbell10 commented on tools/codegen/gen.py:
```diff
@@ -123,6 +123,11 @@ def concatMap(func: Callable[[T], Sequence[S]], xs: Sequence[T]) -> Iterator[S]:
+def escape_string(s: str) -> str:
+    s = s.replace('\\', '\\\\')
+    s = s.replace('"', '\\"')
+    return s
```
Good idea, thanks.

peterbell10 committed:
Update on "codegen: Allow string arguments to have defaults"


Fixes #43944

Note that the codegen doesn't use a proper parser so, in the same way as with lists, the string `, ` cannot appear in defaults or it will be interpreted as a splitting point between arguments.

[ghstack-poisoned]
